### DISCUSSION TITLE
best way to measure similarity

### DISCUSSION THREAD

**Discussion by nadongjin** [leondgarse/Keras_insightface#67] [Q&A]
Hello
Your kind explanation is the best.
Recently, there is one thing I am most curious about.
What is the best way to measure similarity in your recognition model?
For example, I calculate as follows.
Ask for your opinion whether this is correct.

image1 ---------->recognition engine--------> 512 vector (a1_vector)
image2----------->recognition engine---------> 512 vector (a2_vector)

a1_l2=a1_vector/np.linalg.norm(a1_vector)
a2_l2=a2_vector/np.linalg.norm(a2_vector)

cos_similarity=np.sum(a1_l2 * a2_l2)
===================================================
Finally, I have about 5,000 classes of data set for Asian adults and children.
It has 40 to 70 images per class.
I can give it to you if you need it.
===================================================

**Comment by leondgarse** [comment]
- You method just works, but `np.sum(...)` can only process one by one. My mostly used is like:
  ```py
  from sklearn.preprocessing import normalize
  aa = np.random.uniform(size=[5, 512])
  bb = np.random.uniform(size=[3, 512])
  
  # Normalize
  eaa = normalize(aa)  # Same with `eaa = aa / np.linalg.norm(aa, axis=-1, keepdims=True)`
  ebb = normalize(bb)
  print((eaa ** 2).sum(-1), (ebb ** 2).sum(-1))  # Or `print(np.linalg.norm(eaa, axis=-1), np.linalg.norm(ebb, axis=-1))`
  # [1. 1. 1. 1. 1.] [1. 1. 1.]
  
  # Cosine distance.
  dists = np.dot(eaa, ebb.T)
  # For the first line, it's similarity between `eaa[0]` with `ebb[0], ebb[1], ebb[2]`.
  print(dists)
  # [[0.76494142 0.75154709 0.74995526]
  #  [0.7513076  0.76081367 0.73874   ]
  #  [0.75749455 0.74173085 0.74128313]
  #  [0.76652466 0.73447405 0.75479567]
  #  [0.76880897 0.76184213 0.77380811]]
  ```
- If your test dataset is consists of aligned face images, you may use `./eval_folder.py -d {DATA_PATH} -m {BASIC_MODEL.h5}` for evaluating accuracy, or create your own test bin file with additional `-B {BIN_FILE.bin}`. More detail can be found in that script.
- I can't ask for your dataset, as it may not legal, Lol. Actually seldom do I use custom dataset. It's still `MS1MV3` my mostly used one.

**Reply by nadongjin** [reply]
If you look at the report of insightface's nist frvt 1:1, the threshold value is as follows.
I am curious about is that if the cosine similarity method is used, it is normal for the threshold value to be set between [-1,1].
It seems to be calculated by value-(distance), my guess.
I can't find an exact calculation for insightface's measure of similarity.
I think there is something hidden here that I don't know.
[A high value indicates a high degree of similarity.]
![image](https://user-images.githubusercontent.com/14832440/151359539-b15d7d62-650a-4659-afba-2e8b3724f668.png)

**Reply by leondgarse** [reply]
I havn't check that code, maybe it's `euclidean dist`? As `euc_dist = np.sqrt(2 - 2 * cos_dist)` and `cos_dist = (2 - euc_dist ** 2) / 2`, so for `euc_dist = 1.284760`, it's `cos_dist = 0.174696`.
```py
from sklearn.preprocessing import normalize
aa = np.random.uniform(size=[5, 512])
bb = np.random.uniform(size=[3, 512])
# Normalize
eaa = normalize(aa)
ebb = normalize(bb)

cos_dists = np.dot(eaa, ebb.T)
euc_dists = np.sqrt(((np.expand_dims(eaa, 1) - np.expand_dims(ebb, 0)) ** 2).sum(-1))
print(np.allclose(np.sqrt(2 - 2 * cos_dists), euc_dists))
# True
```
But for `euc_dist`, it's a smaller value meaning more similarity...

**Reply by leondgarse** [reply]
Looks like they all using their own measurement, I'm totally unfamiliar with these.
![Selection_441](https://user-images.githubusercontent.com/5744524/151503446-faaacfe5-fc39-4c33-8715-e10f65ea8561.png)

**Reply by nadongjin** [reply]
Thank you for your answer.
I measured it as follows.
In 512 embedding, I measured as below code.
The results were similar insightface's Threshold value
How do you think?
[In fact, it looks like a slight improvement.]
# simple code
def sum_cosine_similaty(vector1,vector2):
    emb1 = vector1 / np.linalg.norm(vector1)
    emb2 = vector2 / np.linalg.norm(vector2)
    emb3=vector1[:256]/np.linalg.norm(vector1[:256])
    emb4=vector2[:256]/np.linalg.norm(vector2[:256])
    emb5 = vector1[256:] / np.linalg.norm(vector1[256:])
    emb6 = vector2[256:] / np.linalg.norm(vector2[256:])
    sum1=np.sum(emb1*emb2)  #all cosine similarity
    sum2=np.sum(emb3*emb4)  # front :256 cosine similarity
    sum3=np.sum(emb5*emb6)  # back 256: cosine similarity
    return sum1+sum2+sum3

**Reply by leondgarse** [reply]
Ya, that looks reasonable when comparing faces with masks. I think it may work.
